pd/src/app.rs

```diff
@@ -443,24 +452,63 @@ impl App {
+                // delegations in this block have not yet been committed, we have to combine
+                // the delegations in pending_block with the ones already committed to the
+                // state. otherwise the delegations committed in the epoch threshold block
+                // would be lost.
+                let mut committed_delegation_changes =
+                    state.delegation_changes(prev_epoch.index).await?;
+                for (id_key, delta) in &pending_block.lock().unwrap().delegation_changes {
+                    let _ = *committed_delegation_changes
+                        .entry(id_key.clone())
+                        .and_modify(|change| *change += delta)
+                        .or_insert(*delta);
+                }
+
                 for current_rate in &current_rates {
                     let identity_key = current_rate.identity_key.clone();
 
                     let funding_streams = state.funding_streams(identity_key.clone()).await?;
 
                     let next_rate = current_rate.next(&next_base_rate, funding_streams);
 
-                    let delegation_token_supply = state
+                    // TODO: if a validator isn't part of the consensus set, should we ignore them
+                    // and not update their rates?
+                    let delegation_delta = committed_delegation_changes
+                        .get(&identity_key)
+                        .unwrap_or(&0i64);
+
+                    // update staking token supply
+                    let unbonded_amount =
+                        current_rate.unbonded_amount(delegation_delta.abs() as u64);
+                    if *delegation_delta > 0 {
+                        // delegation: subtract the unbonded amount from the staking token supply
+                        staking_token_supply -= unbonded_amount;
+                    } else {
+                        // undelegation: add the unbonded amount to the staking token supply
+                        staking_token_supply += unbonded_amount;
+                    }
+
+                    let mut delegation_token_supply = state
                         .asset_lookup(identity_key.delegation_token().id().encode_to_vec())
                         .await?
                         .map(|info| info.total_supply)
                         .unwrap_or(0);
 
-                    // TODO: we should process all of the delegations and undelegations in the
-                    // epoch here
-                    //
-                    // TODO: if a validator isn't part of the consensus set, should we ignore them
-                    // and not update their rates?
+                    delegation_token_supply =
+                        (delegation_token_supply as i64 + delegation_delta) as u64;
+
+                    // update the delegation token supply
+                    // TODO: should we use a method which panics on integer overflow here?
```

Comment:
Is this comment supposed to attach to the line above?  I checked if there were any helpful methods, and there's going to be a `checked_add_signed` [in the future](https://doc.rust-lang.org/stable/std/primitive.u64.html#method.checked_add_signed) but it's not stable now.

Since there's already a branch on the sign of delta above, maybe it would be cleanest to merge the delegation token supply and staking token supply changes together, like
```rust
let delegation_amount = delegation_delta.abs() as u64;
let unbonded_amount = current_rate.unbonded_amount(delegation_amount);

let mut delegation_token_supply = /* ... */

if *delegation_delta > 0 {
    // net delegation: ...
    staking_token_supply = staking_token_supply.checked_sub(unbonded_amount).unwrap();
    delegation_token_supply = delegation_token_supply.checked_add(delegation_amount).unwrap();
} else {
    // net undelegation: ...
    staking_token_supply = staking_token_supply.checked_add(unbonded_amount).unwrap();
    delegation_token_supply = delegation_token_supply.checked_sub(delegation_amount).unwrap();
}
```